API for getting a deterministic key for a Compilation

jaredpar opened the pull request:
The `Compilation` type is fully deterministic meaning that given the same inputs
(`SyntaxTree`, `CompilationOptions`, etc ...) it will produce the same output. By 
that it means it will produce byte for byte equivalent binaries and the same 
diagnostics. This is extremely valuable in infrastructure because it allows 
for caching and opens the door for efficient distributed processing.

At the moment this is hard to leverage because consumers can't determine if 
two `Compilation` instances are equivalent for the purposes of determinism. 
Customers have to resort to hand written comparisons which requires a fairly 
intimate knowledge of the compiler (for example knowing what does and does not
impact determinism). Such solutions are not version tolerant; every time the 
compiler adds a new property that impacts determinism the solution must be 
updated. Even when proper equality checks are in place this does not help 
distributed computing where equality must be decided across different processes.

The motivation here is to provide an API that returns a string based key for a
given `Compilation` such that for two equivalent `Compilation` instances the
key will also be equivalent. This will allow customers who want to leverage 
deterministic caching and communication to have a simple, and transmittable,
way to full describe the `Compilation` they are processing.


Remaining items: 
- [x] File issue tracking API review 
- [x] Write lots more tests once general structure of change is agreed on

Follow up Items: 
- #57940
- #57941

RikkiGibson commented on src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs:
```diff
@@ -4090,6 +4091,8 @@ void writeValue(string key, string value)
             }
         }
 
+        internal override DeterministicKeyBuilder CreateDetermisiticKeyBuilder() => new CSharpDeterministicKeyBuilder();
```
`Determisitic` -> `Deterministic` :)

jaredpar replied:
Well at least I deterministically am terrible at spelling

RikkiGibson commented on src/Compilers/CSharp/Portable/Compilation/CSharpDeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,65 @@
+
+        protected override void WriteCompilationOptionsCore(JsonWriter writer, CompilationOptions options)
+        {
+            if (options is not CSharpCompilationOptions csharpOptions)
```
maybe someday.. https://github.com/dotnet/csharplang/issues/1328

RikkiGibson commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,476 @@
+
+        protected void WriteFileName(JsonWriter writer, string name, string? filePath, DeterministicKeyOptions options)
+        {
+            if (0 != (options & DeterministicKeyOptions.IgnorePaths))
```
nit: prefer constant value on the right

RikkiGibson commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,476 @@
+            EmitOptions? emitOptions = null,
+            DeterministicKeyOptions options = default)
+        {
+            ensureNotDefault(ref additionalTexts);
```
`array = array.NullToEmpty()`?

RikkiGibson commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,476 @@
+                writer.WriteArrayStart();
+                foreach (var analyzer in analyzers)
+                {
+                    WriteType(writer, analyzer.GetType());
```
Why do we need to include the names of individual analyzers and generators rather than the list of analyzer/generator assemblies? Also, should we include MVIDs of the assemblies?

jaredpar replied:
I think the assemblies make sense from a command line compilation perspective but this can also be used at an API level. It is possible to provide individual analyzers / generators at this level that are not the full set in the containing assembly. Hence using the assembly alone isn't enough. 

Will add the MVID though to `WriteType`

RikkiGibson commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,476 @@
+            {
+                writer.WriteKey("toolsVersions");
+                writer.WriteObjectStart();
+                if (0 != (options & DeterministicKeyOptions.IgnoreToolVersions))
```
nit: constant 0 should be on the right. Unless we are making a style decision that we put the 0 on the left for these kinds of flag checks?

RikkiGibson commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,476 @@
+                            }
+                            else
+                            {
+                                throw new InvalidOperationException();
```
Is this a TODO or do we just not support netmodules here? Should we have a comment to that effect?

jaredpar replied:
Yes I need to fix this up before merge.

RikkiGibson commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,476 @@
+            }
+            else
+            {
+                throw new InvalidOperationException();
```
Shouldn't we also be able to write the key of a compilation reference? This is the case when the IDE creates the compilation, right? Is that a TODO here or requires some design consideration to know what to do here?

CyrusNajmabadi replied:
this is an interesting case.  would be happy if you solved it.  if you don't though, we'll just filter out those sorts of metadata references and then mix-in our own checksums for them into the final result.

RikkiGibson commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,476 @@
+            writer.WriteKey("extensions");
+            writer.WriteObjectStart();
+
+            WriteType(writer, "syntaxTreeOptionsProvider", options.SyntaxTreeOptionsProvider?.GetType());
```
What determines these values? Concerned that it's going to be difficult to get consistency between the IDE and batch compilations with these.

jaredpar replied:
These should be generally the same between the IDE and Compiler. If they are not though we should consider adding an `IgnoreExtensions` flag.

RikkiGibson reviewed (commented):
The general approach seems fine. Had a lot of small questions. Very cool progress.

Therzok commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDeterministicKeyBuilder.vb:
```diff
@@ -0,0 +1,74 @@
+                    writer.Write("name", import.Name)
+                    writer.Write("isXml", import.IsXmlClause)
+                    writer.WriteObjectEnd()
+                Next
```
Missing WriteArrayEnd

jaredpar replied:
Thanks, fixed!

CyrusNajmabadi commented on determinism.md:
```diff
@@ -0,0 +1,171 @@
++           ImmutableArray<DiagnosticAnalyzer> analyzers = default,
++           ImmutableArray<ISourceGenerator> generators = default,
++           EmitOptions? emitOptions = null,
++           DeterministicKeyOptions options = DeterministicKeyOptions.Default)
```
can we update to have hte static entrypoint?

jaredpar replied:
I've been trying to think on what the best `static` entry point would be. The implementation of the key can't use `CompilationOptions` for example it needs to use values off of `CSharpCompilationOptions`. One of the advantages of having `Compilation.GetDeterministicKey` is the implementation was in the CSharp or Basic versions which knew what type to cast `CompilationOptions` too. 

Once there is a `static` entry point we lose that virtual dispatch a little bit. I can still add it back in the implementation it is just a bit less pretty. I will take a stab at the design and see what everyone thinks but open to early suggestions as well.

jaredpar replied:
Okay updated with a `static` entry point here sketched out.

CyrusNajmabadi reviewed (changes requested):
ugh. pending changes ae the worst.

jaredpar commented on determinism.md:
```diff
@@ -0,0 +1,171 @@
++       public string GetDeterministicKey(
++           ImmutableArray<AdditionalText> additionalTexts = default,
++           ImmutableArray<DiagnosticAnalyzer> analyzers = default,
++           ImmutableArray<ISourceGenerator> generators = default,
```
What was the representation you were expecting to see here?

jaredpar commented on determinism.md:
````diff
@@ -0,0 +1,171 @@
+```json
+{
+  "options": {
+    "outputKind": "ConsoleApplication",
````
Yeah I struggled with that too but JSON world is camelCase so I leaned in here.

CyrusNajmabadi replied:
yup.  when in rome seems appropriate.  no concern here.

jaredpar commented on determinism.md:
```diff
@@ -0,0 +1,171 @@
++       /// <summary>
++       /// Ignore all file paths, but still include file names, in the deterministic key.
++       /// </summary>
++       IgnorePaths = 0b0001,
```
File names are a gray area of determinism. In the strict sense yes they matter because it impacts data written to the PDB, `[CallerFileInfo]` and certain diagnostic messages. A significant amount of customers are likely not concerned with these differences though. 

Consider back when we had caching in our CI system. The specific path that our code was checked out to was not material to our build and test caching. Or stated differently the same commit run through CI but checked out into two different directories always had the same outcome in terms of build failures and tests. Because source paths matter though we had to jump through hoops in our build and caching logic to normalize out the path differences. This API provides a very simple way to say 

> Give me the key that is independent of the location where the code is built from

The reason I kept the names vs. just deleting the paths entirely is that there is one other case that matters for determinism here. Bit of an edge case but consider that the compiler will error on the command line if you pass the same file twice. Keeping the file name means that we catch that case in the key. 

File names have some other tricky points though that we need to consider and account for. For example "aux.txt" is a valid file name on Windows 11 but not Windows 10. So to have a fully correct key we need to encode the OS version into the key.

jaredpar commented on determinism.md:
```diff
@@ -0,0 +1,171 @@
+    "unsafe": false,
+    "topLevelBinderFlags": "None"
+  },
+  "syntaxTrees": [
```
The overall compiler documentation is pretty explicit about ordering mattering but I can re-enforce that notion here for customers.

CyrusNajmabadi replied:
yeah, i definitely don't think it's necessary.  but i prefer putting up warning signs all over the place for people :)

CyrusNajmabadi commented on determinism.md:
```diff
@@ -0,0 +1,174 @@
+    {
++       public static string GetDeterministicKey(
++           CompilationOptions compilationOptions,
++           ImmutableArray<SyntaxTree> syntaxTrees,
```
definitely will talk about this in our meeting today (specifically teh SyntaxTree part).

RikkiGibson replied:
let's make sure to record what suggestion we have here. e.g. are we going to simply make the parameter `ImmutableArray<SourceText>` instead, and make you pass ParseOptions separately.

CyrusNajmabadi replied:
Unfortunately, the compiler api allows each tree to have different parseoptions.  so i think we need some object that encapsulates the three things that are actually used here).

jaredpar replied:
Moving the API discussion to the issue now, made sure this was recorded in the new API on the issue

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,478 @@
+        {
+            if ((options & DeterministicKeyOptions.IgnorePaths) != 0)
+            {
+                filePath = Path.GetFileName(filePath);
```
these apis can throw.  we should doc at the top level that this can be possible.

jaredpar replied:
Leaving this open until we decide on the actual entry points and will add the docs then.

jaredpar replied:
Added  a comment about throwing to the entry point doc.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,478 @@
+            // Note that the file path to the assembly is deliberately not included here. The file path
+            // of the assembly does not contribute to the output of the program.
+            writer.Write("assemblyName", type.Assembly.FullName);
+            writer.Write("mvid", type.Assembly.ManifestModule.ModuleVersionId.ToString());
```
consider an explicit format flag here.  though i'm not sure if that matters for Guid.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,478 @@
+            ImmutableArray<DiagnosticAnalyzer> analyzers = default,
+            ImmutableArray<ISourceGenerator> generators = default,
+            EmitOptions? emitOptions = null,
+            DeterministicKeyOptions options = default)
```
we should make thsi cancellable.  it does expensive work (that does take cancellation tokens).  I do this in my branch.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,478 @@
+                    writer.WriteObjectStart();
+                    WriteFileName(writer, "fileName", additionalText.Path, options);
+                    writer.WriteKey("text");
+                    WriteSourceText(writer, additionalText.GetText());
```
GetText can be expensive (and does IO.  pass cancellation through).

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,478 @@
+            {
+                writer.WriteKey("toolsVersions");
+                writer.WriteObjectStart();
+                if ((options & DeterministicKeyOptions.IgnoreToolVersions) != 0)
+                {
+                    writer.WriteObjectEnd();
+                    return;
+                }
+
+                var compilerVersion = typeof(Compilation).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
+                writer.Write("compilerVersion", compilerVersion);
+
+                var runtimeVersion = typeof(object).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
+                writer.Write("runtimeVersion", runtimeVersion);
+
+                writer.Write("framework", RuntimeInformation.FrameworkDescription);
+                writer.Write("os", RuntimeInformation.OSDescription);
+
```
```suggestion
                if ((options & DeterministicKeyOptions.IgnoreToolVersions) == 0)
                {
                    var compilerVersion = typeof(Compilation).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
                    writer.Write("compilerVersion", compilerVersion);

                    var runtimeVersion = typeof(object).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
                    writer.Write("runtimeVersion", runtimeVersion);

                    writer.Write("framework", RuntimeInformation.FrameworkDescription);
                    writer.Write("os", RuntimeInformation.OSDescription);
                }

```

CyrusNajmabadi replied:
(saves having to think about early exit needing to close things properly.)

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,478 @@
+            writer.WriteObjectStart();
+            WriteFileName(writer, "fileName", syntaxTree.FilePath, options);
+            writer.WriteKey("text");
+            WriteSourceText(writer, syntaxTree.GetText());
```
need cancellation token here.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,478 @@
+            writeSubsystemVersion(writer, options.SubsystemVersion);
+            writer.Write("fileAlignment", options.FileAlignment);
+            writer.Write("highEntropyVirtualAddressSpace", options.HighEntropyVirtualAddressSpace);
+            writer.Write("baseAddress", options.BaseAddress.ToString());
```
this likely needs passing in invariant culture here.  we should likely do an audit of every ToString to make sure it is valid.

jaredpar replied:
Moved to `"G"` which should round trip across cultures. I'm verifying that now.

jaredpar replied:
Nope I was wrong about this.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,478 @@
+            else
+            {
+                writer.Write("deterministic", false);
+                writer.Write("localtime", options.CurrentLocalTime.ToString(CultureInfo.InvariantCulture));
```
interesting.  so if we ask for a deterministic key for a compilatino we haven't passed the .deterministic option to, we are certain to get something not reusable? (not protesting... just interesting :))

jaredpar replied:
Hmm, good point as it's possible to have a case where you have `/deteriministic-`, the same local time but get different assemblies because with `/deterministic-` we effectively `Guid.NewGuid()` the MVID. Need to add an extra field here called `stopBeingNaughty: Guid.NewGuid()`

CyrusNajmabadi commented on src/Compilers/Core/Portable/InternalUtilities/JsonWriter.cs:
```diff
@@ -88,18 +88,56 @@ public void Write(string key, int value)
+        public void Write<T>(string key, T value) where T : struct, Enum
+        {
+            WriteKey(key);
+            Write(value.ToString());
```
i'm on the fence about this.  i feel like perhaps for enums we just emit the integral value.  the primary concern i have is uncertainty if the runtime always gives the same string value (for things like flags enums) when you call tostring across launches.

jaredpar replied:
The `[Flags]` case should be safe as the runtime guarantees the behavior there. From the [doc comments](https://docs.microsoft.com/en-us/dotnet/api/system.enum.tostring?view=net-6.0)

> If the FlagsAttribute is applied and there is a combination of one or more named constants equal to the value of this instance, then the return value is a string containing a delimiter-separated list of the names of the constants.

CyrusNajmabadi commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDeterministicKeyBuilder.vb:
```diff
@@ -0,0 +1,75 @@
+                For Each pair In basicOptions.PreprocessorSymbols
+                    writer.WriteObjectStart()
+                    writer.WriteKey(pair.Key)
+                    writer.Write(pair.Value.ToString())
```
interesting.  this is .ToString on an aribtrary object.  Do we know what forms this will take and if this is safe?

CyrusNajmabadi replied:
this enumerates a dictionary.  that may have different results across rules of hte same program as i don't believe .net guarantees deterministic order.  Consider sorting things like this to ensure we're ok.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,478 @@
+            {
+                writer.WriteObjectStart();
+                writer.Write(kvp.Key, kvp.Value);
+                writer.WriteObjectEnd();
```
this enumerates a dictionary.  that may have different results across rules of hte same program as i don't believe .net guarantees deterministic order.  Consider sorting things like this to ensure we're ok.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,478 @@
+        private void WriteEmitOptions(JsonWriter writer, EmitOptions? options)
+        {
+            writer.WriteObjectStart();
+            if (options is null)
```
consider inverting.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,478 @@
+            {
+                writer.WriteObjectStart();
+                writer.Write(kvp.Key, kvp.Value);
+                writer.WriteObjectEnd();
```
this enumerates a dictionary.  that may have different results across rules of hte same program as i don't believe .net guarantees deterministic order.  Consider sorting things like this to ensure we're ok.

CyrusNajmabadi commented on src/Compilers/Core/RebuildTest/BasicDeterministicKeyBuilderTests.cs:
```diff
@@ -0,0 +1,118 @@
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.PooledObjects;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.VisualBasic.UnitTests;
```
sort :)

CyrusNajmabadi commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDeterministicKeyBuilder.vb:
```diff
@@ -0,0 +1,75 @@
+
+        Public Sub New()
+
+        End Sub
```
this type appears to be a singleton.  Consider making constructors prviate and expose a single static instance.

CyrusNajmabadi replied:
same applies to C#.

CyrusNajmabadi reviewed (changes requested):
Overall this looks great, and just what i need.  Called out some potential gotchas.  But they all seem fixable.

jaredpar commented on src/Compilers/Core/Portable/Compilation/DeterministicKey.cs:
```diff
@@ -0,0 +1,96 @@
+
+namespace Microsoft.CodeAnalysis
+{
+    internal abstract class SyntaxTreeKey
```
What other derivations of this type were you expecting to write? My intuition was to use a `readonly struct` or `sealed class` here but seems like you had other scenarios in mind that I'm missing.

CyrusNajmabadi replied:
In the IDE i implement this by sitting on top of our Document abstraction.  So GetText forwards to our own impl of getting the text.  It could a readonly-struct, but it would need to be passed a Delegate to do the work.  This seemed cleaner (and in in line with all our 'XXXProvider' types.

jaredpar commented on determinism.md:
```diff
@@ -0,0 +1,171 @@
+
+Note: this scenario does require more work as the `Compilation` objects created
+in the IDE and Compiler differ in subtle ways. Having this function though would
+be motivation to clean these up such that they are equal.
```
I believe we have addressed this with the changes you added to my PR (thank you!). This is an area I want to keep a close eye on though particularly during API review. If we see any more I am happy to continue pushing on the perf aspect here.

CyrusNajmabadi replied:
yes.  i'm happy if we have SyntaxTreeKey.

jaredpar commented:
Summary of the changes: 

1. Marked this ready for review now that I feel the major issues have been addressed 
2. Removed the public API. I still intend to make this public but that is a process. Want to get the implementation agreed on now (at lesast the big parts) then have a separate PR where we make this public as a separate PR. That will allow Cyrus to continue his work as well as for us to start submitting bug / perf fixes based on his testing. 
3. Addressed all the existing feedback except reviewing the `ToString` calls to ensure they are invariant. Outside that if you had feedback that isn't addressed please ping me because I missed it (appologies in advance).

RikkiGibson commented on src/Compilers/Core/RebuildTest/DeterministicKeyBuilderTests.cs:
```diff
@@ -312,6 +313,24 @@ JObject getValue(bool deterministic)
         }
 
+        /// <summary>
+        /// Disabling determinism should mean all calls to GetDeteriministicKey return different values.
```
It seems like it would be reasonable to just throw if GetDeterministicKey is called when determinism is disabled.

jaredpar replied:
I considered that but for now decided against it. It's not inherently wrong to ask for a key on a non-deterministic build, we just have to meet the contract that it never compares equal. I also worried that it would create friction points for consumers where they have to keep taking different paths whether a specific option was passed to the compiler. 

I actually started wondering if the opposite was needed. Essentially make the output deterministic even if the options weren't deterministic. Consider the C# IDE case as motivation for this. They are trying to determine equivalence between open / closing the IDE and remote sessions. That shouldn't all fall apart if the customer didn't use deterministic builds. It also seems odd to call `WithDeterminism(true)` before calling the API. Possible better to just ignore that option in the output.

RikkiGibson replied:
It seems like something we'll figure out the answer for as we set up the consuming side of the API. I agree there is a considerable use case for wanting to determine "equivalence" of compilations even if determinism is disabled. I could see that being its own option or just being the way it always works (if you don't care whether it's literally deterministic, isn't it good enough that the compilation inputs are equivalent except for the things which depend on the moment in time or are straight up random, etc.)

RikkiGibson commented on src/Compilers/Core/RebuildTest/DeterministicKeyBuilderTests.cs:
```diff
@@ -0,0 +1,703 @@
+            Assert.Equal(expected, actual);
+        }
+
+        private protected static void AssertSyntaxTreePathMap(string? expected, CommonCompiler compiler)
```
It wasn't clear to me what the utility was in making some methods `private protected`.

jaredpar replied:
`CommonCompiler` is `internal` so I had to chose between `internal` and `private protected` for these methods. The intent was to be consumed in derived types hence I chose `protected` route

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Compilation/CSharpDeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,66 @@
+
+        private CSharpDeterministicKeyBuilder()
+        {
+
```
```suggestion
```

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Compilation/CSharpDeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,66 @@
+        {
+            if (options is not CSharpCompilationOptions csharpOptions)
+            {
+                throw new InvalidOperationException();
```
ArgumentException, ideally with name of arg?

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Compilation/CSharpDeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,66 @@
+
+            writer.Write("unsafe", csharpOptions.AllowUnsafe);
+            writer.Write("topLevelBinderFlags", csharpOptions.TopLevelBinderFlags);
+            writer.WriteKey("globalUsings");
```
nit: just `usings`.  That's what the option is called.  i think it woudl be good to keep in sync.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Compilation/CSharpDeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,66 @@
+
+        protected override void WriteParseOptionsCore(JsonWriter writer, ParseOptions parseOptions)
+        {
+            if (parseOptions is not CSharpParseOptions csharpOptions)
```
make DeterministicKeyBuilder  generic on TParseOptions and this can be moved to base type and derived types can always get the type they expect.

jaredpar replied:
I experimented with this and the end result is that either: 

1. The `public` API point becomes generic because the generic ends up propogating there through `CreateDeterministicBuilder` 
2. Introduce another level in the hierarchy which is non-generic that we can continue to expose

jaredpar replied:
I chased this idea for a little while and I don't think the result makes the code cleaner. The pattern I ended up with is essentially the following: 

```c#
internal abstract class DeterministicKeyBuilder
{

    protected abstract WriteParseOptions(JsonWriter writer, ParseOptions parseOptions);
}


internal abstract class DeterministicKeyBuilder<TParseOptions, TCompilationOptions> : DeterministicKeyBuilder
    where TParseOptions : ParseOptions
    where TCompilationOptions : CompilationOptions
{

    protected override void WriteParseOptions(JsonWriter writer, ParseOptions parseOptions)
    {
        writer.WriteObjectStart();
        if (parseOptions is not TParseOptions derivedParseOptions)
        {
            throw new ArgumentException(null, nameof(parseOptions));
        }

        WriteParseOptionsCore(writer, derivedParseOptions);
        writer.WriteObjectEnd();
    }
}
```

The non-generic `DeterministicKeyBuilder` is necessary because `CompilationOptions` can't expose a generic type here. Hence you end up with the exact same pattern we were trying to avoid by adding another level in the hierarchy. Didn't feel like a great trade off. 


Let me know if you think I missed an easier path here.

CyrusNajmabadi replied:
nope!  i'm fine with you doing this.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -123,6 +123,77 @@ public abstract partial class Compilation
+        ///   understand the difference between two compilations which is impacting the deterministic 
+        ///   output
+        /// - It is *not* in a minimal form. If used as a key in say a content addressable storage consumers
+        ///   should first pass it through a strong hashing function.
```
fwiw, this is not going to show up properly in QuickInfo.  Need things like `<list>` and `<item>` unfortunately.

jaredpar replied:
May as well do it right :smile:

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -123,6 +123,77 @@ public abstract partial class Compilation
+        ///   should first pass it through a strong hashing function.
+        ///
+        /// Compilations which do not use the /deterministic option can still use this API but
+        /// the results will change on every invocation.
```
notblocking, but we should discuss this.  i think IDE would prefer we could still get good deterministic keys back in this case.  we could always attempt to workaround this by always setting the 'deterministic' option to true before calling you... but it's a bit fugly.

jaredpar replied:
Yeah I discussed this scenario with @RikkiGibson in another comment. One idea I had was adding an entry to `DeterministicKeyOptions` to suppress this behavior because for scenarios IDE is considering it is just likely resulting in needless cache misses.

jcouv replied:
Is "the results will change on every invocation" actually the case, or would the result remain the same in most cases? 
The main exception is when `AssemblyVersionAttribute` is used in certain ways which necessarily result in a non-deterministic compilation (see [refout docs](https://github.com/dotnet/roslyn/blob/main/docs/features/refout.md#determinism).

jaredpar replied:
> Is "the results will change on every invocation" actually the case, or would the result remain the same in most cases?

There is a 100% chance the EXE / DLL will change on every invocation. That is because lacking the `/determinism` flag the MVID will be effectively `Guid.NewGuid()`.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+using Roslyn.Utilities;
+using System.Threading;
+using System.Collections.Generic;
+using Microsoft.Cci;
```
sort

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+            {
+                builder.Append(b.ToString("x"));
+            }
+        }
```
pretty large encoding.  consider base64.  (very minor feedback).

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+
+            writer.WriteObjectStart();
+
+            writer.WriteKey("compilation");
```
nit: consider moving all these .WriteKey's into the respective WriteXXX helpers.

jaredpar replied:
Originally I had this but pulled into this form because several `WriteXXX` helpers are used with different key values. Most notably is actually `WriteCompilation` which gets written at the top level as well as a nested reference. Having the `"compilation": ` indirection in the second case seemed odd (willing to be convinced on this point though)

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+            writer.Write(name, builder.ToStringAndFree());
+        }
+
+        protected void WriteType(JsonWriter writer, string key, Type? type)
```
`Type?`?  can a null value actually sneak in?  feels... strange :)

jaredpar replied:
The majority of types used here are allowed to be `null`. I agree it's strange but it's allowed. 🤷 

Consider the concrete example of the `StrongNameProvider`

```c#
        /// <summary>
        /// Provides strong name and signature the source assembly.
        /// Null if assembly signing is not supported.
        /// </summary>
        public StrongNameProvider? StrongNameProvider { get; protected set; }
```

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+            // Note that the file path to the assembly is deliberately not included here. The file path
+            // of the assembly does not contribute to the output of the program.
+            writer.Write("assemblyName", type.Assembly.FullName);
+            writer.Write("mvid", type.Assembly.ManifestModule.ModuleVersionId.ToString());
```
GetGuidValue?

jaredpar replied:
Good catch

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+            writeToolsVersions();
+
+            WriteByteArrayValue(writer, "publicKey", publicKey.AsSpan());
+            writer.WriteKey("options");
```
nit: move into WriteCompilationOptions?

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+                    writer.Write("runtimeVersion", runtimeVersion);
+
+                    writer.Write("framework", RuntimeInformation.FrameworkDescription);
+                    writer.Write("os", RuntimeInformation.OSDescription);
```
nit: 

frameworkDescription
osDescription

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+            writer.WriteObjectStart();
+            WriteByteArrayValue(writer, "checksum", sourceText.GetChecksum().AsSpan());
+            writer.Write("checksumAlgorithm", sourceText.ChecksumAlgorithm);
+            writer.Write("encoding", sourceText.Encoding?.EncodingName);
```
nit: encodingName

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+                                writeModuleMetadata(modules[i]);
+                                writer.WriteObjectEnd();
+                            }
+                            writer.WriteArrayEnd();
```
i'm curious why it's 1, followed by an array?  why not just do a single array?

jaredpar replied:
The vast majority of assemblies are single module. Multi-module assemblies are incredibly rare and not supported at all on .NET Core. I didn't want to complicate the output for that case.

CyrusNajmabadi replied:
well it's a matter of perspective.  now you'll always have the `"secondaryModules": []` piece.  my recomenddation would be that you either `if` things so that isn't added if not needed.  Or just have it be always an array and hten not have the almost-always-pointless `secondaryModules` bit.  NOte: this is a big nit.  As you stated, this is not a format someone can depend on.  So it was just about potentially making the code to generate simpler (as well as reducing junk in the output).  If you prefer this, that's 100% fine.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+                    {
+                        var assemblyDef = peReader.GetAssemblyDefinition();
+                        writer.Write("name", peReader.GetString(assemblyDef.Name));
+                        writer.Write("version", assemblyDef.Version.ToString());
```
nit: audit tostring.

jaredpar replied:
This `ToString` is maddening because the .NET Core version has an `IFormattable` implementation I can use while desktop does not. Further the numbers are all small value `int` which should be naturally invariant but still want to be sure.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+                builder.WriteCompilation(
+                    writer,
+                    compilation.Options,
+                    compilation.SyntaxTrees.SelectAsArray(x => SyntaxTreeKey.Create(x)),
```
```suggestion
                    compilation.SyntaxTrees.SelectAsArray(static x => SyntaxTreeKey.Create(x)),
```

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+            }
+            else
+            {
+                throw new NotSupportedException();
```
nit: message.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+            WriteType(writer, "xmlReferenceResolver", options.XmlReferenceResolver?.GetType());
+            WriteType(writer, "sourceReferenceResolver", options.SourceReferenceResolver?.GetType());
+            WriteType(writer, "strongNameProvider", options.StrongNameProvider?.GetType());
+            WriteType(writer, "assemblyIdentityComparer", options.AssemblyIdentityComparer?.GetType());
```
ah, i see you can get nulls.

CyrusNajmabadi commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,529 @@
+            }
+            else
+            {
+                writer.WriteNull();
```
consider not special casing and just writing out an empty objeect.

jaredpar replied:
Good catch. Rest of the code works this way, missed this one.

CyrusNajmabadi commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDeterministicKeyBuilder.vb:
```diff
@@ -0,0 +1,109 @@
+            For Each import In basicOptions.GlobalImports
+                writer.WriteObjectStart()
+                writer.Write("name", import.Name)
+                writer.Write("isXml", import.IsXmlClause)
```
isXmlClause

CyrusNajmabadi commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDeterministicKeyBuilder.vb:
```diff
@@ -0,0 +1,109 @@
+                ElseIf type = GetType(Double) Then
+                    writer.Write(pair.Key, CType(value, Double).ToString("G"))
+                ElseIf type = GetType(SByte) Then
+                    writer.Write(pair.Key, CType(value, SByte).ToString("G"))
```
:'(

jaredpar replied:
Yeah I couldn't see a better way to do this which was durable across cultures (once I switch `G` to invariant)

jaredpar replied:
Doh, I can just use `IFormattable` and simplify this a lot

CyrusNajmabadi reviewed (approved):
LGTM.  Just ultra pedantic nits.

jcouv commented on src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -1,7 +1,10 @@
 abstract Microsoft.CodeAnalysis.SyntaxTree.GetLineMappings(System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.LineMapping>!
 const Microsoft.CodeAnalysis.WellKnownMemberNames.PrintMembersMethodName = "PrintMembers" -> string!
-Microsoft.CodeAnalysis.Compilation.EmitDifference(Microsoft.CodeAnalysis.Emit.EmitBaseline! baseline, System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.Emit.SemanticEdit>! edits, System.Func<Microsoft.CodeAnalysis.ISymbol!, bool>! isAddedSymbol, System.IO.Stream! metadataStream, System.IO.Stream! ilStream, System.IO.Stream! pdbStream, System.Threading.CancellationToken cancellationTok ...
+Microsoft.CodeAnalysis.DeterministicKeyOptions
```
Was there an API review on public API changes?

jaredpar replied:
I meant to revert this public API when I reverted the other parts. Will fix.

jcouv commented on src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -156,7 +159,6 @@ static Microsoft.CodeAnalysis.IncrementalValueProviderExtensions.WithComparer<TS
 static Microsoft.CodeAnalysis.Operations.OperationExtensions.GetFunctionPointerSignature(this Microsoft.CodeAnalysis.Operations.IFunctionPointerInvocationOperation! functionPointer) -> Microsoft.CodeAnalysis.IMethodSymbol!
 virtual Microsoft.CodeAnalysis.Diagnostics.AnalyzerReference.GetGenerators(string! language) -> System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.ISourceGenerator!>
 virtual Microsoft.CodeAnalysis.Diagnostics.AnalyzerReference.GetGeneratorsForAllLanguages() -> System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.ISourceGenerator!>
-abstract Microsoft.CodeAnalysis.Compilation.GetUsedAssemblyReferences(System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> System.Collections.Immutable.ImmutableArray<Microsoft.CodeAnalysis.MetadataReference!>
```
Why does it look like some public APIs were removed (`GetUsedAssemblyReferences` here and `EmitDifference` above at line 4)?
If it's uninentional, did CI properly catch that?

jaredpar replied:
Have no idea how this changed. Maybe a bad merge? Will fix.

cston commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDeterministicKeyBuilder.vb:
```diff
@@ -0,0 +1,109 @@
+                ElseIf type = GetType(Boolean) Then
+                    writer.Write(pair.Key, CType(value, Boolean))
+                ElseIf type = GetType(DateTime) Then
+                    writer.Write(pair.Key, CType(value, DateTime).ToString("G"))
```
Consider extracting a `Shared Function ConvertToString(value As Object) As String` helper method.

jaredpar replied:
`object.ToString()` though is parmeterless hence I can't force invariant formatting. If there was a way to do this I'd love to hear it cause I don't particularly like this code .

jaredpar commented:
> Just ultra pedantic nits.

Keep the pedantic comments coming. Details are really subtle on this feature so no comment is too pedantic here. 😄

chsienki commented on src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:
```diff
@@ -1649,68 +1651,21 @@ protected virtual CultureInfo Culture
         {
-            var key = CreateDeterminismKey(args, rawArgs, baseDirectory, parser);
-            var filePath = Path.Combine(args.OutputDirectory, args.OutputFileName + ".key");
+            var key = compilation.GetDeterministicKey(additionalTexts, analyzers, generators, pathMap, emitOptions);
```
Generators can change the code they produce based on the values passed in through analyzer config files, so I think we need to include those too.

chsienki replied:
This potentially overlaps with some of the work I mentioned here: https://github.com/dotnet/roslyn/pull/57177 

We may consider adding equality for the `AnalyzerConfigOptionsProvider`, in which case we wouldn't need to consider the input analyzer config files themselves, but instead the parsed values according to the rules of editorconfig, which would allow for functionally equivalent edits to still be treated as equal.

jaredpar commented:
Huh, the tests are failing in CI but passing locally ...

jaredpar commented:
Huh, looks like this code already found a determinism bug. The RESX file content in our Roslyn.VisualStudio.DiagnosticsWindow build are changing between builds. This doesn't show up in the end DLL / PDB but is showing up in the `/additionalFiles` entry

jaredpar commented:
Ah it's because they're using full paths in the XLF RESX files 

```xml
  <data name="400" type="System.Resources.ResXFileRef, System.Windows.Forms">
    <value>C:\Users\jaredpar\code\roslyn2\src\VisualStudio\VisualStudioDiagnosticsToolWindow\Resources\Package.ico;System.Drawing.Icon, System.Drawing, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a</value>
  </data>
  <data name="301" type="System.Resources.ResXFileRef, System.Windows.Forms">
    <value>C:\Users\jaredpar\code\roslyn2\src\VisualStudio\VisualStudioDiagnosticsToolWindow\Resources\Images.png;System.Drawing.Bitmap, System.Drawing, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a</value>
  </data>
```

That means once we do the `subst` build on a different directory the values are changing.

jaredpar commented:
@CyrusNajmabadi in debugging through some of the failures here for real compilations I discovered the JSON files are indeed enormous. It's likely realizing your fears around this causing a lot of unnecessary big string allocations. I'm already planning on doing the work to de-dupe parse trees but that won't be enough. After this PR is merged I think we'll need to start a second where we add an option to return the compressed output and avoid the big string allocation.

chsienki commented:
@jaredpar I guess I'm surprised that we're considering the analyzers+generators etc at all when building the compilation key, given that they exist outside of the compilation. Are we conflating two concepts here: one of the `Compilation` object itself and the conceptual 'Compilation' as a set of inputs that produce an output on disk?

If we are, I would think that instead we should have each 'thing' (compilation, analyzer collection, pathmap) be able to produce a determinism key (effectively a fancy hashcode) which can then be combined together to determine if a conceptual compilation will produce the same outputs.

The fact that you could ask individual elements for their key seems useful and also means the `compilation` object itself doesn't have to own any responsibility/awareness for the external stuff that may change and how to emit a key for it.

chsienki commented:
I could imagine some sort of interface like `IDeterminable` that indicates a compiler object supports determinism. You could pass a keybuilder to it, which allows it to append its information. That way we could have multiple output strategies: one where we just calculate a hash on the fly for most operations, and one where we actually create the JSON file for human readability and debugging.

Then if you want to just compare two `Compilation` objects, you get them to emit their keys, and you don't have to consider anything else. You could do the same for a eg. a collection of generators to see if we can re-use previous state.

jaredpar commented:
@chsienki 

> Are we conflating two concepts here: one of the Compilation object itself and the conceptual 'Compilation' as a set of inputs that produce an output on disk?

For this to be useful I think you need to be able to support both scenarios. Consider that the IDE team needs to decide if two project builds are equivalent. That means they need the root `Compilation` as well as the additional files, analyzers, generators, etc ... to have a full comparison. 

The API is designed so that you can pick the elements to include the key. Essentially you can decide if you want the core `Compilation` only or if you want to add in the other elements that can impact the output. 

One big lesson from this work that is hard to convey is that determinism is not a yes / no type of feature. It actually has a good deal of gray area in it. What does and does not matter is going to differ between consumers and I want the API to support all of them.

> I could imagine some sort of interface like IDeterminable that indicates a compiler object supports determinism

I discussed the interface approach on the associated issue. I think it's a valid approach to the problem but I worry that it ends up spreading out the logic quite a bit. I think there is value in having all of the logic for "this does or does not impact deterministic outpu" in the same file. It's a complicated subject and it's nice to be able to view it all in a single place vs. spreading it out amongst the code base. 

I intentionally pulled the public API changes out of this PR because I wanted to have a separate discussion about them. The change itself is already pretty involved and wanted to get the logic / tests in then potentially iterate on how we expose it after. 

> Then if you want to just compare two Compilation objects, you get them to emit their keys, and you don't have to consider anything else. You could do the same for a eg. a collection of generators to see if we can re-use previous state.

I can't really see a lot of scenarios where we can **only** consider the `Compilation`. In order for the key to be useful it really needs to consider the other inputs. At least in the cases I've considered. I'm very interested in scenarios where other data may not be relevant.

jcouv commented on src/Compilers/Core/RebuildTest/BasicDeterministicKeyBuilderTests.cs:
```diff
@@ -0,0 +1,226 @@
+using Microsoft.CodeAnalysis.VisualBasic.UnitTests;
+using System.Collections.Generic;
+
+namespace Microsoft.CodeAnalysis.Rebuild.UnitTests
```
nit: consider file-scoped namespace for new test files

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyOptions.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Microsoft.CodeAnalysis
+{
+    internal enum DeterministicKeyOptions
```
Looks like these are flags. Consider adding the attribute

jcouv commented on eng/test-determinism.ps1:
```diff
@@ -24,8 +24,12 @@ if ($help) {
-  "Microsoft.CodeAnalysis.EditorFeatures2.UnitTests.dll"
+  "Microsoft.CodeAnalysis.EditorFeatures2.UnitTests.dll",
+
+  # TOOD: file XLF additional files bug before merge
```
nit: FWIW, I've added "TODO2" to the list of words checked by the CI leg. Unlike "PROTOTYPE" it is checked in all branches. It's handy to track this sort of thing.

jaredpar replied:
Got the issue filed for this.

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,532 @@
+        }
+
+        protected static void WriteByteArrayValue(JsonWriter writer, string name, ReadOnlySpan<byte> value)
+        {
```
method could be simplified to `writer.Write(name, EncodeByteArrayValue(value))`, avoiding duplication with above `EncodeByteArrayValue`?

jaredpar replied:
Doing that though would introduce an unnecessary intermediate `string` allocation.

jcouv commented on src/Compilers/Core/RebuildTest/DeterministicKeyBuilderTests.cs:
```diff
@@ -0,0 +1,727 @@
+        {
+            var checksum = text.GetChecksum();
+            var builder = PooledStringBuilder.GetInstance();
+            DeterministicKeyBuilder.EncodeByteArrayValue(checksum.AsSpan(), builder);
```
Could use the `EncodeByteArrayValue` method that takes a single ReadOnlySpan parameter (and no builder)

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,532 @@
+            return builder.ToStringAndFree();
+        }
+
+        internal static void EncodeByteArrayValue(ReadOnlySpan<byte> value, StringBuilder builder)
```
Could be made inlined after callers are adjusted to call single-parameter method above?

jaredpar replied:
The preference is to use this method which writes directly into the existing `StringBuilder`. The other version is not preferred and really only used in test code. Can move that to test helper if you think it makse the code more readable.

jcouv commented on src/Compilers/Core/Portable/Compilation/Compilation.cs:
```diff
@@ -123,6 +123,101 @@ public abstract partial class Compilation
+        /// </list>
+        /// This API can throw exceptions in a few cases like invalid file paths.
+        /// </remarks>
+        internal static string GetDeterministicKey(
```
private?

jaredpar replied:
Why `private`? This is intended to eventually be the `public` entry point. Kept it `internal` for easier test and prototype access while we vet out the implementation.

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKey.cs:
```diff
@@ -0,0 +1,105 @@
+        }
+    }
+
+    internal static class DeterministicKey
```
nit: is there a benefit of separating `DeterministicKey` and `DeterministicKeyBuilder`? I'd be tempted to move those static APIs over into `DeterministicKeyBuilder`.

jcouv replied:
@jaredpar Did you have a though on this?

jaredpar replied:
Presently I like the separation because the `DeterministicKey` API is completely stateless. It's a `static class` which has all inputs which are effectively immutable data and returns a key. It fits with the theme of determinism. The builder APIs though are stateful, involve hierachies and non-static types. 

I thought about merging them but then we end up having to expose `abstract class` instances to customers for types they can never instantiate and for which there are no instance methods available. Felt wrong.

jcouv replied:
Alright. Thanks

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,532 @@
+            WriteType(writer, type);
+        }
+
+        protected void WriteType(JsonWriter writer, Type? type)
```
Do we have a test that shows what this outputs?

jaredpar replied:
Hmm, good catch. Thought I had added that but apparently missed that case.

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,532 @@
+            WriteCompilation(writer, compilationOptions, syntaxTrees, references, publicKey, pathMap, options, cancellationToken);
+            writer.WriteKey("additionalTexts");
+            writeAdditionalTexts();
+            writer.WriteKey("analyzers");
```
Do we have tests with non-empty analyzers? 
Same question for generators.

jaredpar replied:
Fixed

jcouv commented on src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs:
```diff
@@ -1080,6 +1077,11 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger? errorLogger, Cancella
                     }
                 }
 
+                if (Arguments.ParseOptions.Features.ContainsKey("debug-determinism"))
```
This is existing code, but seems to be lacking test coverage (can't find "debug-determinism" anywhere else)

jaredpar replied:
This feature flag is used in the determinism PR check and it was added ~5 years ago when I first started pushing on determinism in the compiler. This flag just writes out the deterministic key file next to the binary. When the deterministic check fails it allows the developer to diff the key files and determine if the compilation was given different inputs. If there are different inputs then determinism failed due to a build system bug, if it's the same inputs then it's a compiler bug. 

This has saved me and others massive amounts of time over the years.

jcouv replied:
I'm not questioning this feature flag :-) My observation is just about lack of test coverage.

jaredpar replied:
Gotcha. Yes this code was added ~5 years ago and had no test coverage. That is why no tests failed as a result of this. Can try and add some test coverage here (generally this path is awkward to test which is likely why it wasn't tested in the first place).

jaredpar replied:
Spent some time this morning adding a test case here. I focused on the fact that it is hard to write tests that cover this code path. Hence spent some time factoring out a bit of code to make it easier. It looks promising and should be as easy as the following when I am done.

```c#
var (compiler, fileSystem) = CSharpCompilerUtil.CreateForNetCoreApp("test.cs");
compiler.Run();
````

The change is a bit invasive though. Do you want me to keep building on this PR or finish of this PR and submit the refactor + test of this code path as a sep change? Fine with either.

jcouv replied:
Thanks for looking into it. Since this is not central to current PR, let's punt.

jaredpar replied:
I went ahead and put it in the current PR. When working on the change I found a bug in this path that I wanted to fix. I put it as a separate commit to make it easier to review.

jcouv replied:
Just curious, where was the bug?

jaredpar replied:
There were actually two bugs: 

1. The placement of the feature flag check meant that it only ran if the compilation had analyzers or generators. 
2. The original `Arguments.EmitOptions` was being used instead of the local `emitOptions` which meant it didn't have items like the `pdbPath` set.

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,532 @@
+        {
+            if (sourceText is null)
+            {
+                return;
```
Should we do `writer.WriteNull();` here (similar to how `WriteType` handles nulls)?

jaredpar replied:
Good catch that was a bug. Added a test to expose the bug and fixed this up.

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,532 @@
+                            writeModuleMetadata(modules[0]);
+                            writer.WriteKey("secondaryModules");
+                            writer.WriteArrayStart();
+                            for (var i = 1; i < modules.Length; i++)
```
nit: foreach?

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,532 @@
+                writer.WriteKey("properties");
+                writeMetadataReferenceProperties(writer, reference.Properties);
+
+                void writeModuleMetadata(ModuleMetadata moduleMetadata)
```
nit: We try to put all local functions at the end of the method, if possible, to avoid interrupting the main flow of the code

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,532 @@
+            {
+                var compilation = compilationReference.Compilation;
+                var builder = compilation.Options.CreateDeterministicKeyBuilder();
+                builder.WriteCompilation(
```
Do we have a test for this case?

jaredpar replied:
We do now :smile:

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,532 @@
+            DeterministicKeyOptions deterministicKeyOptions)
+        {
+            writer.WriteObjectStart();
+            if (options is not null)
```
Should we write null instead of `{ }`?

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,532 @@
+            writer.WriteObjectStart();
+            if (options is not null)
+            {
+                writer.Write("emitMetadataOnly", options.EmitMetadataOnly);
```
As we add new options (or more generally extend the compilation object model), is there anything we can do to remind us to update this code?
An entry mentioning determinism key may be useful in the compiler test plan. Maybe comments in the relevant types might help too?

jaredpar replied:
The unit tests are designed to help guard against this because they effectively count the number of properties on the critical types. If a new property is added then the tests will fail and the test has a comment about making sure to update this representation. 

If you have suggestions on how to make these tests more robust I'm def open to them.

jaredpar replied:
See the `VerifyUpToDate` tests

jcouv replied:
Thanks. `VerifyUpToDate` seems a good approach

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,532 @@
+                writer.Write("includePrivateMembers", options.IncludePrivateMembers);
+                writer.WriteKey("instrumentationKinds");
+                writer.WriteArrayStart();
+                foreach (var kind in options.InstrumentationKinds)
```
Could `InstrumentationKinds` be default?

jcouv commented on src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,532 @@
+            // of compilations this is roughly the equivalent of recording the compiler version but it 
+            // could differ when customers host the compiler via the API.
+            writer.WriteKey("extensions");
+            writer.WriteObjectStart();
```
nit: consider adding a local scope to mirror the nesting

jcouv commented on src/Compilers/CSharp/Portable/Compilation/CSharpDeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,65 @@
+
+            // Even though tools like the command line parser don't explicitly order the symbols 
+            // here the order doesn't actually impact determinism.
+            foreach (var symbol in csharpOptions.PreprocessorSymbols.OrderBy(StringComparer.Ordinal))
```
Would it be better to keep the original order then (easier to compare with command line)?

jaredpar replied:
The original order isn't recoverable because the command line uses non-ordered collections. I looked into changing it to use ordered collections but it is a pretty significant change. It doesn't really positively impact any other part of the code hence thought it better to just normalize it here. 

Comparing with the original command line isn't really a scenario. Instead it's generally comparing the deterministic key file from two different command lines.

jcouv commented on src/Compilers/CSharp/Portable/Compilation/CSharpDeterministicKeyBuilder.cs:
```diff
@@ -0,0 +1,65 @@
+            writer.WriteArrayEnd();
+        }
+
+        protected override void WriteParseOptionsCore(JsonWriter writer, ParseOptions parseOptions)
```
Should parseOptions be marked as nullable? (I see that the VB override handles null)

jaredpar replied:
It is a bit of a tricky case. The only case where you get `null` is when grabbing the `ParseOptions` from `VisualBasicCompilation`. That only happens when you construct it via the API, it will always have one for command line compilations.

Decided to keep this non-null and add an explicit check in the VB code to guard against it.

jcouv replied:
That's fair. 
nit: Consider adding assertion or comment in C# side

jcouv commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDeterministicKeyBuilder.vb:
```diff
@@ -0,0 +1,94 @@
+                writer.WriteKey(pair.Key)
+
+                Dim type = value.GetType()
+                If type = GetType(String) Then
```
Do we have test coverage for all cases here? I saw tests with string and with integer values, but not boolean.

jaredpar replied:
Yep was missing `bool` and `float`. fixed.

jcouv commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs:
```diff
@@ -54,6 +54,7 @@ internal class DefaultAnalyzerAssemblyLoader : AnalyzerAssemblyLoader
                "System.Runtime.CompilerServices.Unsafe",
                "System.Runtime.Extensions",
                "System.Runtime.InteropServices",
+               "System.Runtime.InteropServices.RuntimeInformation",
```
nit: I didn't follow this. We can chat tomorrow.

jcouv reviewed (commented):
Done with review pass (iteration 49)

jaredpar commented:
thanks for the feedback @jcouv. Please let me know if I missed any of your comments.

jcouv commented on src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDeterministicKeyBuilder.vb:
```diff
@@ -0,0 +1,100 @@
+            If basicOptions.ParseOptions IsNot Nothing Then
+                WriteParseOptions(writer, basicOptions.ParseOptions)
+            Else
+                writer.WriteObjectStart()
```
Should this be `WriteNull` instead?

jcouv replied:
This is the last open question

jcouv replied:
I saw you marked this as resolved, but I'm not sure what the resolution was. Could you clarify?

jaredpar replied:
I swear I flipped that to `WriteNull` as I marked it resolved but it's not showing up so clearly I didn't. Let me try a second time 😦

jaredpar replied:
Oh I did add it but i did not push 🤦

jcouv reviewed (commented):
Done with review pass (iteration 52)

jcouv commented on src/Compilers/Test/Core/Extensions.cs:
```diff
@@ -0,0 +1,25 @@
+
+namespace Roslyn.Test.Utilities
+{
+    internal static class Extensions
```
>Extensions [](http://example.com/codeflow?start=26&length=10)

nit: CommonCompilerExtensions

jcouv commented on src/Compilers/Test/Core/TestableCompiler.cs:
```diff
@@ -0,0 +1,215 @@
+
+namespace Roslyn.Test.Utilities
+{
+    internal record struct TestableCompilerFile(string FilePath, TestableFile TestableFile)
```
nit: we haven't really used records in the compiler codebase yet because we didn't really need the structural equality part. But this in tests, so I think I'm okay with using records for brevity.

jaredpar replied:
There are some decently heavy use of `record` in the rebuild code base. But you are right I don't need structural equality here so reverting to plain `struct`

jcouv commented on src/Compilers/Test/Core/TestableCompiler.cs:
```diff
@@ -0,0 +1,215 @@
+        Manual,
+    }
+
+    internal sealed class TestableCompiler
```
nit: consider adding xml doc

jcouv commented on src/Compilers/Test/Core/TestableCompiler.cs:
```diff
@@ -0,0 +1,215 @@
+        internal TestableFileSystem FileSystem { get; }
+        internal BuildPaths BuildPaths { get; }
+
+        internal TestableCompiler(CommonCompiler compiler, TestableFileSystem fileSystem, BuildPaths buildPaths)
```
>fileSystem [](http://example.com/codeflow?start=78&length=10)

Why are we passing in a `fileSystem` if it has to be the same as `compiler.FileSystem`? Can't we just use `compiler.FileSystem` then, and remove the extra parameter?

jaredpar replied:
The problem with the property is that it attaches *after* command line arguments are parsed where the ctor parameter has a chance to thread it through. It's long been a problem with this setup and one that I'm slowly working my waay trhough.

jcouv reviewed (approved):
LGTM Thanks (iteration 54)

jcouv reviewed (approved):
LGTM Thanks (iteration 55)
